File: java/revenj-servlet/src/main/java/org/revenj/server/commands/crud/Create.java
package org.revenj.server.commands.crud;

import org.revenj.patterns.*;
import org.revenj.security.PermissionManager;
import org.revenj.server.CommandResult;
import org.revenj.server.ServerCommand;
import org.revenj.serialization.Serialization;

import java.io.IOException;
import java.security.Principal;
import java.util.Optional;

public final class Create implements ServerCommand {

	private final DomainModel domainModel;
	private final PermissionManager permissions;

	public Create(
			DomainModel domainModel,
			PermissionManager permissions) {
		this.domainModel = domainModel;
		this.permissions = permissions;
	}

	public static final class Argument<TFormat> {
		public String Name;
		public TFormat Data;
		public Boolean ReturnInstance;

		public Argument(String name, TFormat data, Boolean returnInstance) {
			this.Name = name;
			this.Data = data;
			this.ReturnInstance = returnInstance;
		}

		@SuppressWarnings("unused")
		private Argument() {
		}
	}

	@Override
	public <TInput, TOutput> CommandResult<TOutput> execute(
			ServiceLocator locator,
			Serialization<TInput> input,
			Serialization<TOutput> output,
			TInput data,
			Principal principal) {
		Argument<TInput> arg;
		try {
			arg = input.deserialize(data, Argument.class, data.getClass());
		} catch (IOException e) {
			return CommandResult.badRequest(e.getMessage());
		}
		Optional<Class<?>> manifest = domainModel.find(arg.Name);
		if (!manifest.isPresent()) {
			return CommandResult.badRequest("Unable to find specified domain object: " + arg.Name);
		}
		if (arg.Data == null) {
			return CommandResult.badRequest("Data to create not specified.");
		}
		if (!AggregateRoot.class.isAssignableFrom(manifest.get())) {
			return CommandResult.badRequest("Specified type is not an aggregate root: " + arg.Name);
		}
		if (!permissions.canAccess(manifest.get(), principal)) {
			return CommandResult.forbidden(arg.Name);
		}
		AggregateRoot instance;
		try {
			instance = (AggregateRoot) input.deserialize(manifest.get(), arg.Data);
		} catch (IOException e) {
			return CommandResult.badRequest("Error deserializing provided input for: " + arg.Name + ". Reason: " + e.getMessage());
		}
		PersistableRepository repository;
		try {
			repository = locator.resolve(PersistableRepository.class, manifest.get());
		} catch (ReflectiveOperationException e) {
			return CommandResult.badRequest("Error resolving repository for: " + arg.Name + ". Reason: " + e.getMessage());
		}
		try {
			repository.insert(instance);
			boolean returnInstance = arg.ReturnInstance == null || arg.ReturnInstance;
			return new CommandResult<>(output.serialize(returnInstance ? instance : instance.getURI()), "Object created", 201);
		} catch (IOException e) {
			return CommandResult.badRequest(e.getMessage());
		}
	}
}

File: scala/revenj-akka/src/main/scala/net/revenj/server/commands/crud/Create.scala
package net.revenj.server.commands.crud

import net.revenj.patterns.{AggregateRoot, DomainModel, PersistableRepository, ServiceLocator}
import net.revenj.serialization.Serialization
import net.revenj.server.commands.crud.Create.Argument
import net.revenj.server.{CommandResult, ServerCommand}

import scala.concurrent.Future

class Create(domainModel: DomainModel) extends ServerCommand {

  override def execute[TInput, TOutput](
    locator: ServiceLocator,
    input: Serialization[TInput],
    output: Serialization[TOutput],
    data: TInput): Future[CommandResult[TOutput]] = {

    val arg = input.deserializeRuntime(data, classOf[Argument[TInput]], data.getClass)
    lazy val manifest = domainModel.find(arg.get.Name)
    if (!arg.isSuccess) {
      CommandResult.badRequest(arg.failed)
    } else if (manifest.isEmpty) {
      CommandResult.badRequest(s"Unable to find specified domain object: ${arg.get.Name}")
    } else if (arg.get.Data == null) {
      CommandResult.badRequest("Data to create not specified.")
    } else if (!classOf[AggregateRoot].isAssignableFrom(manifest.get)) {
      CommandResult.badRequest(s"Specified type is not an aggregate root: ${arg.get.Name}")
    } else {
      val instance = input.deserializeRuntime[AggregateRoot](arg.get.Data, manifest.get)
      if (!instance.isSuccess) {
        CommandResult.badRequest(s"Error deserializing provided input for: ${arg.get.Name}. Reason: ${instance.failed.get.getMessage}")
      } else {
        val tryRepository = locator.resolve(classOf[PersistableRepository[AggregateRoot]], manifest.get)
        if (!tryRepository.isSuccess) {
          CommandResult.badRequest(s"Error resolving repository for: ${arg.get.Name}. Reason: ${tryRepository.failed.get.getMessage}")
        } else {
          import scala.concurrent.ExecutionContext.Implicits.global
          tryRepository.get.insert(instance.get).map { uri =>
            val returnInstance = arg.get.ReturnInstance.getOrElse(true)
            val response = output.serializeRuntime(if (returnInstance) instance.get else uri)
            if (response.isSuccess) {
              CommandResult[TOutput](Some(response.get), "Object created", 201)
            } else {
              CommandResult[TOutput](None, response.failed.get.getMessage, 500)
            }
          }
        }
      }
    }
  }
}

object Create {

  case class Argument[TFormat](Name: String, Data: TFormat, ReturnInstance: Option[Boolean])

}